src/builder/QuarkBuilder.sol

```diff
@@ -122,39 +217,44 @@ contract QuarkBuilder {
         Payment calldata payment,
         ChainAccounts[] calldata chainAccountsList
     ) external pure returns (BuilderResult memory) {
-        // TODO: Input validation: Check that arrays are equal length (e.g. chainIds and maxCosts in Payment)
-        AssetPositions[] paymentAssetAccounts;
-        // Get transfer and payment token
-        // TODO: implement filterAccounts
-        transferAssetAccounts = filterAccounts(assetSymbol, chainAccountsList);
+        if (payment.chainIds.length != payment.maxCosts.length) {
+            revert InvalidInput();
+        }
+
+        AssetPositionsWithChainId[] transferAssetPositions = filterAssetPositionsForSymbol(assetSymbol, chainAccountsList);
+        AssetPositionsWithChainId[] paymentAssetPositions;
         if (payment.isToken) {
-            paymentAssetAccounts = filterAccounts(payment.currency, chainAccountsList);
+            paymentAssetPositions = filterAssetPositionsForSymbol(payment.currency, chainAccountsList);
         }
 
         // INSUFFICIENT_FUNDS
-        // There are not enough funds to fulfill the transfer.
-        // aggregate amount of asset on every chain < transfer amount
-        uint256 transferAssetBalance;
-        for (uint i = 0; i < transferAssetAccounts.length; ++i) {
-            for (uint j = 0; j < transferAssetAccounts[i].accountBalances.length; ++j) {
-                transferAssetBalance += transferAssetAccounts[i].accountBalances[j].balance;
+        // There are not enough aggregate funds on all chains to fulfill the transfer.
+        uint256 aggregateTransferAssetBalance;
+        for (uint i = 0; i < transferAssetPositions.length; ++i) {
+            for (uint j = 0; j < transferAssetPositions[i].accountBalances.length; ++j) {
+                aggregateTransferAssetBalance += transferAssetPositions[i].accountBalances[j].balance;
             }
         }
-        if (transferAssetBalance < amount) {
+        if (aggregateTransferAssetBalance < amount) {
             revert InsufficientFunds();
         }
 
+        // TODO: Pay with bridged USDC?
         // MAX_COST_TOO_HIGH
-        // There are not enough funds to satisfy the total max payment cost, after transferring.
+        // There are not enough funds on each chain to satisfy the total max payment cost, after transferring.
```

Comment:
Nit: `There is at least one chain that does not have sufficient payment assets to cover the maxCost for that chain.`

Reply:
Much clearer, thanks